fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/EtagCacheRequestHandler.java

```diff
@@ -0,0 +1,136 @@
+import com.yahoo.bard.webservice.web.responseprocessors.ResponseProcessor;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * Request handler to check the cache for a matching eTag.
+ * <p>
+ * If a query is cached, the etag of the query is retrieved from the cache and is injected into query request header
+ * <p>
+ * Etag-based cache mechanism relies entirely on Druid's etags to determine a cache hit or miss.
+ */
+public class EtagCacheRequestHandler extends BaseDataRequestHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EtagCacheRequestHandler.class);
+    private static final MetricRegistry REGISTRY = MetricRegistryFactory.getRegistry();
+    private static final Meter CACHE_MISSES = REGISTRY.meter("queries.meter.cache.misses");
+    private static final Meter CACHE_REQUESTS = REGISTRY.meter("queries.meter.cache.total");
+
+    protected final DataRequestHandler next;
+    protected final TupleDataCache<String, String, String> dataCache;
+
+    /**
+     * Build a Cache request handler.
+     *
+     * @param next  The next handler in the chain
+     * @param dataCache  The cache instance
+     * @param mapper  The mapper for all JSON processing
+     */
+    public EtagCacheRequestHandler(
+            @NotNull DataRequestHandler next,
+            @NotNull TupleDataCache<String, String, String> dataCache,
+            @NotNull ObjectMapper mapper
+    ) {
+        super(mapper);
+        this.next = next;
+        this.dataCache = dataCache;
+    }
+
+    @Override
+    public boolean handleRequest(
+            RequestContext context,
+            DataApiRequest request,
+            DruidAggregationQuery<?> druidQuery,
+            ResponseProcessor response
+    ) {
+        ResponseProcessor nextResponse = response;
+
+        String cacheKey = null;
+        try {
+            if (context.isReadCache()) {
+                cacheKey = getKey(druidQuery);
+                final TupleDataCache.DataEntry<String, String , String> cacheEntry = dataCache.get(cacheKey);
+                CACHE_REQUESTS.mark(1);
+
+                String eTagInRequest = DruidJsonRequestContentKeys.ETAG.getName();
+                if (cacheEntry != null) { // Current query is in data cache
+                    // Insert "If-None-Match" header into RequestContext; the value is etag of the corresponding cache
+                    context.getHeaders().putSingle(
+                            eTagInRequest,
+                            mapper.readTree(cacheEntry.getValue()).get(DruidJsonResponseContentKeys.ETAG.getName())
+                                    .asText()
+                            );
```

Comment:
Suggestion: I don't like the wrapping here.  how about putting the .get down to align with the .asText